### PR TITLE
[Issue 8154] [Python client] Expose schema version (of writerSchema) in Message

### LINKED ISSUE #8176: [Python Client] Pass schema version of writer schema to decoded message
This is a prerequisite for implementing Issue #8154

### PR DESCRIPTION
**Pull request by shiv4289**
Fixes #8176

### CODE REVIEW

**Comment by BewareMyPower** [comment]
The cpp tests failed because your code was not well formatted. You should use `clang-format-5.0` to format code.

Also you could `git apply` following diff:

```diff
diff --git a/pulsar-client-cpp/include/pulsar/c/message.h b/pulsar-client-cpp/include/pulsar/c/message.h
index 8b4b612..f54d025 100644
--- a/pulsar-client-cpp/include/pulsar/c/message.h
+++ b/pulsar-client-cpp/include/pulsar/c/message.h
@@ -202,7 +202,7 @@ PULSAR_PUBLIC int pulsar_message_get_redelivery_count(pulsar_message_t *message)
 
 PULSAR_PUBLIC int pulsar_message_has_schema_version(pulsar_message_t *message);
 
-PULSAR_PUBLIC const char* pulsar_message_get_schemaVersion(pulsar_message_t *message);
+PULSAR_PUBLIC const char *pulsar_message_get_schemaVersion(pulsar_message_t *message);
 
 PULSAR_PUBLIC void pulsar_message_set_schema_version(pulsar_message_t *message, const char *schemaVersion);
 
diff --git a/pulsar-client-cpp/lib/MessageImpl.cc b/pulsar-client-cpp/lib/MessageImpl.cc
index 8acb959..5d1edbf 100644
--- a/pulsar-client-cpp/lib/MessageImpl.cc
+++ b/pulsar-client-cpp/lib/MessageImpl.cc
@@ -98,7 +98,7 @@ void MessageImpl::setRedeliveryCount(int count) { redeliveryCount_ = count; }
 
 bool MessageImpl::hasSchemaVersion() const { return metadata.has_schema_version(); }
 
-void MessageImpl::setSchemaVersion(const std::string& schemaVersion) { schemaVersion_ =  &schemaVersion; }
+void MessageImpl::setSchemaVersion(const std::string& schemaVersion) { schemaVersion_ = &schemaVersion; }
 
 const std::string& MessageImpl::getSchemaVersion() const { return metadata.schema_version(); }
 
```

**Comment by sijie** [comment]
@shiv4289 overall looks good to me. @BewareMyPower can probably help you add the test case.

**Comment by shiv4289** [comment]
> 
> Also you could `git apply` following diff:
> 

Applied the suggested diff and updated PR.

**Comment by BewareMyPower** [comment]
I saw some unit tests failed by `./build/retry.sh: not found`. It looks like the `retry.sh` was somehow deleted.  Could you give some help? @sijie @codelipenghui

**Comment by shiv4289** [comment]
This error (./build/retry.sh: No such file or directory) is seen in the log in every run after a successful build. Retries are not helping.

```
2020-10-06T05:47:01.5179092Z [100%] Built target main
2020-10-06T05:47:03.7971812Z ##[group]Run ./build/retry.sh pulsar-client-cpp/docker-tests.sh
2020-10-06T05:47:03.7972605Z [36;1m./build/retry.sh pulsar-client-cpp/docker-tests.sh[0m
2020-10-06T05:47:03.8200137Z shell: /bin/bash -e {0}
2020-10-06T05:47:03.8200403Z env:
2020-10-06T05:47:03.8200948Z   JAVA_HOME_8.0.265_x64: /opt/hostedtoolcache/jdk/8.0.265/x64
2020-10-06T05:47:03.8201457Z   JAVA_HOME: /opt/hostedtoolcache/jdk/8.0.265/x64
2020-10-06T05:47:03.8201954Z   JAVA_HOME_8_0_265_X64: /opt/hostedtoolcache/jdk/8.0.265/x64
2020-10-06T05:47:03.8202371Z ##[endgroup]
2020-10-06T05:47:03.8287415Z /home/runner/work/_temp/bc84e325-9808-450b-afae-966a2d0cd8a7.sh: line 1: ./build/retry.sh: No such file or directory
2020-10-06T05:47:03.8294667Z ##[error]Process completed with exit code 127.
2020-10-06T05:47:03.8362423Z Post job cleanup.
2020-10-06T05:47:04.1905004Z Post job cleanup.
```

**Comment by wolfstudy** [comment]
@BewareMyPower @shiv4289 Can help check why ci fails, and make sure that we can merge in 2.6.2.

**Comment by BewareMyPower** [comment]
@wolfstudy It's a rarely seen problem, each CI fails with the same reason:

> /home/runner/work/_temp/e35ee3c2-fd25-4719-bc75-efd7a38be497.sh: line 1: ./build/retry.sh: No such file or directory

It looks like the environment to run CI tests is totally broken. Rerun the CI makes no difference. The only solution I can think of is just close it and open a new PR.

**Comment by sijie** [comment]
@BewareMyPower It is not the CI problem. `build/retry.sh` is a script introduced recently. The GitHub workflows are using this file. But it seems that branch `schema-version` is based on the old branch which doesn't have the retry.sh file. I have re-based Shiv's branch to the latest master. Let's see.

File: pulsar-client-cpp/python/schema_test.py [hunk 1]
```diff
@@ -354,6 +354,32 @@ class Example(Record):
...
+        client = pulsar.Client(self.serviceUrl)
+        producer = client.create_producer(
+                        'my-avro-python-schema-version-topic',
+                        schema=AvroSchema(Example))
+
+        r = Example(a=1, b=2)
+        producer.send(r)
+
+        consumer = client.subscribe('my-avro-python-schema-version-topic', 'sub-1',
+                                    schema=AvroSchema(Example))
```
[the remark this answers is not in the document]
Comment (shiv4289) [reply pulsar-client-cpp/python/schema_test.py]: Done, thanks for spotting that.

File: pulsar-client-cpp/python/schema_test.py [hunk 2]
```diff
@@ -354,6 +354,32 @@ class Example(Record):
...
+        
+        r = Example(a=1, b=2)
+        producer.send(r)
+
+        msg = consumer.receive()
+
+        self.assertIsNotNone(msg.schema_version())
+
+        self.assertEqual(0, int.from_bytes(msg.schema_version().encode(), byteorder='big'))
```
Comment (BewareMyPower) [inline pulsar-client-cpp/python/schema_test.py:377]: `int.from_bytes` is a Python3 method, but CI environment use Python2 to test. You need to use another method to convert bytes to int.

Comment (shiv4289) [reply pulsar-client-cpp/python/schema_test.py]: Fixed as suggested. Thank you!

**Review by jiazhai — APPROVED** [review]

File: pulsar-client-cpp/include/pulsar/c/message.h [hunk 3]
```diff
@@ -200,6 +200,12 @@ PULSAR_PUBLIC const char *pulsar_message_get_topic_name(pulsar_message_t *messag
 
 PULSAR_PUBLIC int pulsar_message_get_redelivery_count(pulsar_message_t *message);
 
+PULSAR_PUBLIC int pulsar_message_has_schema_version(pulsar_message_t *message);
```
Comment (migmit) [inline pulsar-client-cpp/include/pulsar/c/message.h:203]: Am I correct that this (and other two) functions were never implemented?

Comment (BewareMyPower) [reply pulsar-client-cpp/include/pulsar/c/message.h:203]: Yeah, you can open a PR to add the missed implementation.

**Pull request merged**